larpandoracontent/LArControlFlow/StitchingCosmicRayMergingTool.cc

```diff
@@ -764,35 +674,58 @@ void StitchingCosmicRayMergingTool::CalculateX0(const PfoToLArTPCMap &pfoToLArTP
 
             // Calculate X0 for the closest pair of vertices
             LArPointingCluster::Vertex pointingVertex1, pointingVertex2;
-
             try
             {
                 LArStitchingHelper::GetClosestVertices(*pLArTPC1, *pLArTPC2, pointingCluster1, pointingCluster2,
                     pointingVertex1, pointingVertex2);
 
-                PfoToPointingVertexMap::iterator pfoToPointingVertexMapIter1(pfoToPointingVertexMap.find(pPfo1));
-                if (pfoToPointingVertexMapIter1 == pfoToPointingVertexMap.end())
+                PfoToPointingVertexMatrix::iterator pfoToPointingVertexMatrixIter1(pfoToPointingVertexMatrix.find(pPfo1));
+                if (pfoToPointingVertexMatrixIter1 == pfoToPointingVertexMatrix.end())
                 {
-                    (void) pfoToPointingVertexMap.insert(PfoToPointingVertexMap::value_type(pPfo1, pointingVertex1));
+                    PfoToPointingVertexMap pfoToPointingVertexMap({{pPfo2, pointingVertex1}});
+                    (void) pfoToPointingVertexMatrix.insert(PfoToPointingVertexMatrix::value_type(pPfo1, pfoToPointingVertexMap));
                 }
                 else
                 {
-                    if ((pfoToPointingVertexMapIter1->second.GetPosition() - pointingVertex1.GetPosition()).GetMagnitude() > std::numeric_limits<float>::epsilon())
-                        throw StatusCodeException(STATUS_CODE_FAILURE);;
+                    PfoToPointingVertexMap pfoToPointingVertexMap(pfoToPointingVertexMatrixIter1->second);
```

Comment:
This is a really good point. Essentially yes, it should be a reference. At this stage the code is going through pairs that can be stitched and that give a contribution to the X0. We know that we are then going to shift the pfos. You have to shift each pfo in context of a pair/TPC boundary so that you know which way to shift the pfo (the direction is determined by comparing the vertex to the TPC boundary). So, at this stage the stitching vertex is recorded for each match. So, for a pfo you can have two matches and so i record the vertex for each match. 

The fact that i did not use a reference means that a repeated match is not recorded, and this should break the code at a later stage because it would try and look up something in the map which does not exist. But it doesn't break and I think we have been lucky/unlucky (depending on your point of view). So it doesn't break because when shifting the order of the pairs considered is exactly the same as the order in the x0 calculation. But, pfos are only shifted once which means that if a pfo had two matches we would only ever access the first vertex identified which is the one that is inserted in the map. So we never actually try and access the second vertex. 

Obviously I will change this to a reference. 

I think this code is really messy and I think there must be a much simpler way to do this. I just need to give it some thought. I think if we are really clever we may be able to avoid a std::hash as it seems we are storing more info then we actually need...